server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```
@@ -3,51 +3,251 @@
+            // already started work
+            if (next != null && runState.compareAndSet(EnqueuedState.WAITING, EnqueuedState.RUNNING)) {
+                // We successfully set state to running, and should remove the item we just peeked at
+                EnqueuedStreamOperation actualNext = operations.poll();
+                Assert.eq(next, "next", actualNext, "actualNext");
+
+                // Run the new item
+                next.run();
+            }
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            // Avoid starting more work
+            runState.set(EnqueuedState.CLOSED);
+
+            // Safely inform the client that an error happened
+            GrpcUtil.safelyError(responseObserver, errorTransformer.transform(t));
+
+            // If the objecttype connection was opened, close it and refuse later calls from it
+            if (messageStream != null) {
+                closeMessageStream();
+            }
+
+            // Don't attempt to run additional work
+            operations.clear();
+
+            // Release the object itself
+            release();
+        }
+
+        private void closeMessageStream() {
+            try {
+                messageStream.onClose();
+            } catch (Exception ignored) {
+                // ignore errors from closing the plugin
+            }
+        }
+
+        @Override
+        public void onCompleted() {
+            // Don't finalize until we've processed earlier messages
+            runOrEnqueue(() -> {
+                runState.set(EnqueuedState.CLOSED);
+                // Respond by closing the stream - note that closing here allows the server plugin to respond to earlier
+                // messages without error, but those responses will be ignored by the client.
+                GrpcUtil.safelyComplete(responseObserver);
+
+                // Let the server plugin know that the remote end has closed
+                closeMessageStream();
```

It's possible that a client starts a grpc call and then immediately onCompletes it - meaning, messageStream might be null?

---

True. The `closeMessageStream` call is wrapped in a "ignore Exception" though, which would guard against at least getting logs or exceptions about that (though by the same token, onError could skip the null check). Agree they should be consistent, will wrap this one in a null check unless you have another preference.

---

That's fine